File: exercism/ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 3
end

class Squares

  def initialize(n)
    @n = n
  end

  def sum_of_squares
    power_sum(2, 1)
  end

  def square_of_sum
    power_sum(1, 2)
  end

  def difference
    square_of_sum - sum_of_squares
  end

  private
  def power_sum(inner, outer)
    (1 .. @n).reduce(0) { |sum, x| sum + x ** inner } ** outer
  end
end

File: exercism/ocaml/difference-of-squares/difference_of_squares.ml
let square_of_sum (n: int) : int =
  (n + 1) * (n + 1) * n * n / 4

let sum_of_squares (n: int) : int =
  (2 * n + 1) * (n + 1) * n / 6

let difference_of_squares (n: int) : int =
  square_of_sum n - sum_of_squares n
